File: java/MajorityElement.java
/*
 * Filename: MajorityElement.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Thu 18 Jun 2015 11:05:15 AM CST
 */

public class Solution {
	public int majorityElement(int[] nums) {
		int idx = 0;
		int cnt = 1;

		for (int i = 0; i < nums.length; i++) {
			if (nums[i] == nums[idx]) {
				cnt++;
			} else {
				cnt--;
			} 

			if (cnt == 0) {
				idx = i;
				cnt = 1;
			}
		}

		return nums[idx];
	}
}

File: cpp/majority_element.cc
/*
 * Filename: majority_element.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 14 Aug 2015 12:04:55 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		int majorityElement(vector<int>& nums) {
			int mode = nums[0];
			int cnt = 1;

			for (auto i = 1; i < nums.size(); i++)
			{
				if (nums[i] == mode)
				{
					cnt++;
				}
				else
				{
					cnt--;

					if (cnt == 0)
					{
						mode = nums[i];
						cnt = 1;
					}
				}
			}

			return mode;
		}
};
